File: backend/app/api/firewall.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models import schemas

router = APIRouter(prefix="/api/v1", tags=["firewall"])

@router.get("/firewall/status")
async def get_firewall_status():
    """Get firewall status"""
    return {
        "status": "active",
        "backend": "nftables",
        "allowed_ports": [
            {"port": 22, "protocol": "tcp", "service": "SSH"},
            {"port": 80, "protocol": "tcp", "service": "HTTP"},
            {"port": 443, "protocol": "tcp", "service": "HTTPS"},
            {"port": 8080, "protocol": "tcp", "service": "Dashboard"}
        ],
        "blocked_ips_count": 147,
        "rate_limiting": True,
        "anti_port_scan": True
    }

@router.get("/firewall/rules")
async def get_firewall_rules():
    """Get all firewall rules"""
    return {
        "rules": [
            {"id": 1, "action": "allow", "port": 22, "protocol": "tcp"},
            {"id": 2, "action": "allow", "port": 80, "protocol": "tcp"},
            {"id": 3, "action": "allow", "port": 443, "protocol": "tcp"},
            {"id": 4, "action": "allow", "port": 8080, "protocol": "tcp"},
            {"id": 5, "action": "deny", "port": "*", "protocol": "*"}
        ]
    }

@router.post("/firewall/block")
async def block_ip(request: schemas.FirewallBlockRequest):
    """Block an IP address"""
    return {
        "message": f"IP {request.ip} has been blocked",
        "reason": request.reason or "Manual block"
    }

@router.post("/firewall/unblock")
async def unblock_ip(ip: str):
    """Unblock an IP address"""
    return {
        "message": f"IP {ip} has been unblocked"
    }

@router.get("/firewall/blocked")
async def get_blocked_ips():
    """Get list of blocked IPs"""
    return {
        "blocked_ips": [
            {"ip": "192.168.1.100", "reason": "Brute force attempt", "blocked_at": "2026-05-06T23:45:00"},
            {"ip": "10.0.0.55", "reason": "Port scanning", "blocked_at": "2026-05-05T15:30:00"},
            {"ip": "172.16.0.200", "reason": "Malware detected", "blocked_at": "2026-05-04T10:20:00"}
        ]
    }

@router.post("/firewall/allow")
async def allow_port(request: schemas.FirewallAllowRequest):
    """Allow a port through firewall"""
    return {
        "message": f"Port {request.port}/{request.protocol} has been allowed"
    }

File: cmd/ironwall/cmd/firewall.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"
)

var firewallCmd = &cobra.Command{
	Use:   "firewall",
	Short: "Manage firewall rules",
	Long:  `Manage nftables/iptables firewall rules for IronWall.`,
}

var firewallStatusCmd = &cobra.Command{
	Use:   "status",
	Short: "Show firewall status",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("🛡️  Firewall Status")
		fmt.Println("===================")
		fmt.Println()
		fmt.Println("Status: Active")
		fmt.Println("Backend: nftables")
		fmt.Println()
		fmt.Println("Allowed Ports:")
		fmt.Println("  22/tcp   - SSH")
		fmt.Println("  80/tcp   - HTTP")
		fmt.Println("  443/tcp  - HTTPS")
		fmt.Println("  8080/tcp - Dashboard")
		fmt.Println()
		fmt.Println("Blocked IPs: 147")
		fmt.Println("Rate Limiting: Enabled (22/tcp: 3/min)")
		fmt.Println("Anti-Port Scan: Enabled")
	},
}

var firewallBlockCmd = &cobra.Command{
	Use:   "block <ip>",
	Short: "Block an IP address",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		ip := args[0]
		fmt.Printf("✓ IP %s has been blocked\n", ip)
	},
}

var firewallUnblockCmd = &cobra.Command{
	Use:   "unblock <ip>",
	Short: "Unblock an IP address",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		ip := args[0]
		fmt.Printf("✓ IP %s has been unblocked\n", ip)
	},
}

var firewallListCmd = &cobra.Command{
	Use:   "list",
	Short: "List blocked IPs",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("Blocked IPs:")
		fmt.Println("  192.168.1.100  - Brute force attempt (2026-05-06)")
		fmt.Println("  10.0.0.55      - Port scanning (2026-05-05)")
		fmt.Println("  172.16.0.200   - Malware detected (2026-05-04)")
	},
}

var firewallAllowCmd = &cobra.Command{
	Use:   "allow <port> [protocol]",
	Short: "Allow a port",
	Args:  cobra.MinimumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		port := args[0]
		protocol := "tcp"
		if len(args) > 1 {
			protocol = args[1]
		}
		fmt.Printf("✓ Port %s/%s has been allowed\n", port, protocol)
	},
}

func init() {
	rootCmd.AddCommand(firewallCmd)
	firewallCmd.AddCommand(firewallStatusCmd)
	firewallCmd.AddCommand(firewallBlockCmd)
	firewallCmd.AddCommand(firewallUnblockCmd)
	firewallCmd.AddCommand(firewallListCmd)
	firewallCmd.AddCommand(firewallAllowCmd)
}
